guides/hack/51-experimental-features/02-read-only.md

```diff
@@ -0,0 +1,213 @@
+  readonly Bar $bar,
+) : void {
+  $x->mut_prop = $bar; // error, $bar is readonly but the prop is mutable
+  $x->ro_prop = $bar; // ok
+}
+```
+
+### Lambdas and function type signatures
+`readonly` is allowed on inner parameters and return types on function typehints.
+
+``` Hack
+function call(
+    (function(readonly Bar) : readonly Bar) $f,
+    readonly Bar $arg,
+   ) : readonly Bar {
+   return $f($arg);
+}
+```
+
+### Expressions
+`readonly` can appear on expressions.
+
+``` Hack
+function foo(): void {
+  $x = new Foo();
+  $y = readonly $x;
+}
+```
+
+### Functions / Methods
+`readonly` can appear as a modifier on instance methods.
+
+``` Hack
+class C {
+  public readonly function bar(): Bar {
+    return new Bar();
+  }
+  public readonly function foo() : void {
+    $this->prop = 5; // error, $this is readonly.
+  }
+}
+```
+
+#### Other restrictions
+Readonly objects can only call readonly methods.
+
+``` Hack
+class C {
+  public readonly function bar(): Bar {
+    return new Bar();
+  }
+  public readonly function foo() : void {
+    $this->prop = 5; // error, $this is readonly.
+  }
+}
+```
+
+Readonly values can not be passed to a function that takes mutable values.
+
+``` Hack
+function takes_mutable(Foo $x): void {
+  $x->prop = 4;
+}
+
+$z : readonly
+takes_mutable($z); // error, takes_mutable's first parameter
+                   // is mutable, but $z is readonly
+```
+
+And, functions can not return readonly values unless they are marked to return readonly.
+
+``` Hack
+function returns_mutable(readonly Foo $x): Foo {
+  return $x; // error, $x is readonly
+}
+```
+
+### Closures
+A function type can be marked readonly: `readonly function(T1): T`.
+
+## Interactions with [Context and Capabilities](https://docs.hhvm.com/hack/contexts-and-capabilities/available-contexts-and-capabilities)
+If your function only has the `ReadGlobals` capability (i.e. is marked `read_globals` or `policied`) it can only access class static variables if they are wrapped in a readonly expression:
+
+``` Hack
+function read_static()[read_globals]: void {
+  $y = readonly Foo::$bar; // keyword required
+}
+function read_static()[policied]: void {
+  $y = readonly Foo::$bar; // keyword required
+}
+```
+
+### Calling a readonly function
+Calling a function or method that returns readonly requires wrapping the result in a readonly expression.
+
+``` Hack
+function returns_readonly(): readonly Foo {
+  return readonly new Foo();
+}
+
+function test(): void {
+  $x = readonly returns_readonly(); // this is required to call returns_readonly()
+}
+```
+
+## Converting to non-readonly
+Sometimes you may encounter a readonly value that isn’t an object (i.e. a readonly int, due to the deepness property of readonly). In those cases, instead of returning a readonly int, you’ll want a way to tell Hack that the value you have is actually a value type. You can use the function `HH\Readonly\as_mut()` to convert any primitive type from readonly to mutable.
```

Comment:
s/i.e./e.g./

Are objects not supported, or just a very bad idea?

Reply:
I happen to know this. `as_mut()` validates that the argument is a valid type.https://github.com/facebook/hhvm/commit/8114b9f18c28d754fc48b17d64689a57ead44818

Reply:
Adding note that it's usage is only for primitive types and value-type collections of primitive types